cpp/test/sg/fil_test.cu

```diff
@@ -154,31 +159,33 @@ void hard_clipped_bernoulli(
     });
 }
 
+template <typename real_t>
 struct replace_some_floating_with_categorical {
   float* fid_num_cats_d;
   int num_cols;
-  __device__ float operator()(float data, int data_idx)
+  __device__ real_t operator()(real_t data, int data_idx)
   {
-    float fid_num_cats = fid_num_cats_d[data_idx % num_cols];
-    if (fid_num_cats == 0.0f) return data;
+    real_t fid_num_cats = static_cast<real_t>(fid_num_cats_d[data_idx % num_cols]);
```

Comment:
Can we use one way to cast or instantiate types? This line has `static_cast<real_t>` while the next line has `real_t(...)`. If choosing the former (or later, although I prefer the former), we should use AAA style and tag the variable with `auto`

Reply:
Currently, I use `T()` for constant expressions (e.g. `real_t(3)`), and `static_cast<T>()` for non-constant ones (e.g. as in line 168 above).

I replaced `real_t` with `auto` as you suggested.

Do we have any guidelines on using `static_cast<T>()` vs `T()`?

Reply:
We don't have any guidelines, yet, be we should focus effort soon on having a consistent standard for this in our developer guide. I will put that on my plate